src/os/PluggableDBStore.h

```diff
@@ -0,0 +1,172 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2015 SanDisk
+ *
+ * Author: Varada Kari<varada.kari@sandisk.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it.
+ */
+
+/* 
+ * Contains a generic interface to integrate any backend DB in form a shared
+ * object. All needed interfaces should be implemented mentioned in
+ * PluggableDBInterfaces.h and PluggableDBIterator.h in the layer which
+ * interacts with the backend DB. 
+ */
+#ifndef PLUGGABLEDB_STORE_H
+#define PLUGGABLEDB_STORE_H
+
+#include "include/types.h"
+#include "include/buffer.h"
+#include "KeyValueDB.h"
+#include <set>
+#include <map>
+#include <string>
+#include "include/memory.h"
+#include <tr1/memory>
+#include <boost/scoped_ptr.hpp>
+#include <boost/thread/tss.hpp>
+#include <tr1/unordered_map>
+#include <errno.h>
+#include "common/errno.h"
+#include "common/dout.h"
+#include "include/assert.h"
+#include "common/Formatter.h"
+#include "common/ceph_context.h"
+#include "PluggableDBIterator.h"
+#include "PluggableDBInterfaces.h"
+#include <sys/stat.h>
+
+class PerfCounters;
+
+enum {
+  l_PluggableDB_first = 34400,
+  l_PluggableDB_gets,
+  l_PluggableDB_txns,
+  l_PluggableDB_last,
+};
+
+/**
+ * Interface to implement proprietary KeyValue stores
+ */
+class PluggableDBStore : public KeyValueDB {
+  CephContext *cct;
+  PerfCounters *logger;
+  bool init_done;
+  static bool create_if_missing;
+  void *library;
+  int do_open(ostream &out, bool create_if_missing);
+  int getosdid();
+  PluggableDBIterator* DBGetIterator();
+  value_t* DBGetObject(const string& key);
+  int DBInit(const char*, int, map<string, string>*);
+  void DBClose();
+  uint64_t DBGetSize();
+  int DBGetStatfs(struct statfs *buf);
+  int DBSubmitTransaction(map<string, DBOp>& ops);
```

Comment:
All the function pointer named as ptr_* and handles to function pointers are named as DB*. will change them.